Readme
==============================================================================
GRTensorIII Source code, release notes and build procedure

This repo holds the source code for GRTensorIII package for Maple

For installation and usage see the Wiki tab on the GITLAB page
==============================================================================

The code has been restructured to reside in a single Maple module, grtensor.

The Module is defined in build/griii.mpl

The Makefile will run the build/griii.mpl using command line Maple provided
that the Makefile has been edited to have the correct paths. 

The use of a Module/Package in Maple required re-scoping of many of the internal
variables used in GRTensor. Briefly:
- all component info is now in gr_data[<root>, <metric>, etc.]
- spacetime dimension is in Ndim[<metric>]
- active metric is grG_metricName




==============================================================================
GRTensorII 1.xx: Source code release notes and build procedure

August 1999
==============================================================================

These notes refer to the GRTensorII Release 1.xx source code.

GRTensor II is a computer algebra package for performing calculations
in the general area of differential geometry. Its purpose is the
calculation of tensor components on curved spacetimes specified in
terms of a metric or set of basis vectors. The package contains a
library of standard definitions of a large number of commonly used
curvature tensors, as well as the Newman-Penrose formalism. The
standard object libraries are easily expandable by a facility for
defining new tensors. Calculations can be carried out in spaces of
arbitrary dimension, and in multiple spacetimes simultaneously. Though
originally designed for use in the field of general relativity,
GRTensorII is useful in many other fields. 

GRTensor II is not a standalone package, but requires an algebraic
engine. The program was originally developed for MapleV. An earlier
version has been ported to Mathematica under the name GRTensorM.

These notes describe the MapleV version of GRTensorII.

Contents:

1. How to get the source code
2. What you will need to build GRTensorII
3. Building and installing from the source code
   3.1 Building the libraries using 'make'
   3.2 Building the libraries by hand
4. Setting up Maple to run GRTensorII
5. Starting up GRTensorII
6. Reporting bugs/suggestions
7. Submitting patches
8. Notes

More information regarding GRTensor can be obtained from:
  http://grtensor.phy.queensu.ca/

==============================================================================
1. How to get the source code
==============================================================================

HTTP/FTP:

At the moment, source code is not being distributed from the ftp
site, but can be requested from the authors by sending an email to
  <grtensor@astro.queensu.ca>

CVS:

The latest source code, including changes made since the last release,
can be obtained from the GRTensor CVS archive located at:
  :ext:grtensor@grtensor.phy.queensu.ca:/export/home/grtensor/CVS

Local CVS access:
For example, users with an account on grtensor.phy.queensu.ca can obtain
an up-to-date copy of the source:
 setenv CVSROOT /export/home/grtensor/CVS
 cvs checkout grii

Remote CVS access:
At the present time, anonymous CVS access is not available. This means
only users with an account on grtensor.phy.queensu.ca can access
the source from a remote machine.
1. Set up your home directory on grtensor to use ssh. This involves
   creating a .ssh2 directory, putting your public key from the remote
   machine there, and creating an authorisation file referring to your
   public key. Read the ssh documentation for more information.
2. On the remote machine, set the CVSROOT and CVS_RSH variables:
   setenv CVSROOT :ext:grtensor@grtensor.phy.queensu.ca:/export/home/grtensor/CVS
   setenv CVS_RSH ssh
3. You should then be able to check out the source on the remote
   machine as usual:
   cvs checkout grii

CVS software is freely available from:
  http://www.cyclic.com/
Full documentation is included with the distribution. A tutorial is
also available at:
  http://www.csc.calpoly.edu/~dbutler/tutorials/winter96/cvs/

==============================================================================
2. What you will need to build GRTensorII
==============================================================================

Maple:

You will need a copy of the MapleV computer algebra system, Release >= 4.0. 
Unfortunately, this system is not available freely. More information
can be obtained from the Symbolic Computation Group of the University of
Waterloo:

  http://daisy.uwaterloo.ca/home.html

The GRTensor source code is identical for Releases 4 and 5 of
MapleV. In fact, things should also work with at most minor
modifications on older MapleV.3 distributions, though we don't
actively work for R3 backward compatibility anymore or build R3
libraries very often. GRTensorII will not work under MapleV Release 2
because of changes to the way global and local variables were treated.

*** what do we know about student version compatibility?


Other tools:

In order to build the source code distribution, the presence of
certain commonly available tools is assumed: 

  tar, gzip, make, install, sed.

These tools are standard on most Unix platforms and Linux
distributions. Versions compatible with other operating systems are
also available.

If you are using CVS and wish to generate a ChangeLog based on the CVS
archive, the perl script cvs2cl can be used. This script is available from:
  http://www.red-bean.com/~kfogel/cvs2cl.shtml
A copy is also stored in the extras/ directory of the GRTensor source
distribution.  This script requires Perl5 in order to run.

==============================================================================
3. Building and installing from the source code
==============================================================================

The GRTensorII distribution consists of a number of MapleV library
(.m) files and a directory of sample spacetime metric files.

Basically, the build procedure consists of reading all of the *.mpl
files in the maple/grii directory into maple and then saving session
as grii.m. If you can do this by any means you like, you can build
GRTensor. We've also put together some scripts to automate the
procedure which are described here.


3.1 Building the libraries using make:

  Read through the Makefile and make sure that the environment
  variables are set correctly for your system. Type

    make grtensor

  to build the GRTensorII libraries. To install these (including the
  metric directories) on your system, use the command:

    make install


3.2 Building the libraries by hand:

   If you do not have GNU make or some of the tools necessary to build
   GRTensorII using the Makefile, you can install it manually by using
   the scripts in the build/ directory.

  1. Modify the build scripts:

   A number of maple scripts exist in the build/ directory, one for
   each GRTensorII library which can be built. The most important of
   these is the file grii.mpl which specifies the instructions for
   building the main GRTensorII library.

   Edit each of the *.mpl the files in the following way:

   * Near the top, a variable named rootDir is set. Ensure
     that this is pointing to the directory in which the grii source
     code is stored. The directory can be given either as an absolute
     path, or as a relative path from the directory in which you will
     later run the maple command.

   * In the grii.mpl file, the variables grG_Version and grG_Date
     variables are normally set using a sed command which processes
     the Version file (in the root directory of the source tree). If sed
     is not available, set these variables manually.

  2. Build the GRTensorII libraries:

     (1) Open a clean MapleV session using the command
           maple -s
         or
           xmaple -s

     (2) Read one of the script files into maple using the command:
         read <filename>.mpl:

     The maple session should terminate once the file is read, and a
     MapleV library file named
       <filename>.m
     should have been created in the current directory.

     Perform steps (1) and (2) for each of the *.mpl files in the
     build directory.

     The following files should have been created:
        grii.m basislib.m dinvar.m invar.m trigsin.m maple.hdb

  3. Install the libraries:

     Move the library files created in the previous step to a
     directory on your system where you would like the GRTensorII
     package to reside permanently.

  4. Install the metric directory:

     Create a directory on your system where you would like
     GRTensorII metric files to reside. Copy the *.mpl files from the
     metrics/ directory into the directory which you have created.

==============================================================================
4. Setting up Maple to run GRTensorII
==============================================================================

In order to run GRTensorII, you will need to set up MapleV so that it

News
Preserved for history.
Recent info can be found on https://gitlab.com/grtensor/GRTensorIII

==============================================================================
Summary of changes since GRTensorII 1.74:
==============================================================================

Release 1.76 (3 November 1999)

  * Fixes to raising indices for n>5
  * Fixes to automatic determination of index symmetries when indices
    are raised/lowered
  * qload() switches to metric if it has already been loaded.
  * Fixes to clearing of definitions via grundef()
  * Added initzero(), setcmpt(), getsym(), setsym() undocumented hacks.
  * Added expandsqrt(), expandtb() simplification routines.

Release 1.75 (August 1999)

  * Fixes a bug in index ordering in grdisplay() output. (Thanks to
      Jan Gutowski for pointing this out.)
  * Modified user-prompt for interactive functions (eg. makeg()).

Release 1.74 (27 July 1999)

  * Added check for name conflicts in autoAlias().
  * Fixed grtransform() based on source contributed by Edward Huff.

==============================================================================
Changes prior to 1.74:
==============================================================================
173 [27 May 99] dp  Added feature to enter basis vectors as differentials
                    in makeg()
172b[10 Feb 99] dp  Fixed index ordering problem in dispStr3.
                    Fixed support for >6 indices (grF_unassignLoopVars)
                    Fixed grdef() index parser for cbdn, pbdn
                    Fixed autoAlias for R5 & alias 0th order terms
171 [19 Oct 98] dp  Fixed parse errors (for operators and tensors as
                    arguments of functions) in grdef().
                    Modified grsavedef() so save grdef() statements.
170 [31 May 98] dp  Fixed indices on lhs of grdisplay output so that
                    square brackets don't show up in R5 output.
169 [ 6 May 98] dp  Added recognition of derivative and basis
                    indices in grdef consistency checks.
                    Replaced kdelta function with a standard object
                    definition.
168 [15 Mar 98] dp  Added version 5 specific extra arg to `&proc`
167 [24 Feb 98] dp  Add index consistency check in grdef().
                    Fix grundef() to clear assigned objects and all index
                    combinations.
                    Added CMB.
166 [14 Feb 98] dp  Switch reference to `string' data-type to `name'
                    for compatibility with R5.
165 [05 Dec 97] dp  Added grOptionMessageLevel.
164 [04 Nov 97] dp  Fix assignment of grG_metric name in automatic creation
                    of the calc function by grdef().
163 [18 Sep 97] dp  Fix 1-term line-element entry in makeg(), output of
                    grsaveobj() in R3, added initFn.mpl.
162 [16 Sep 97] dp  Comment out `For the xxx metric' in output of grsaveoj()
                    Fix LieD[] operator.
156 [13 Mar 97] dp  Fix greqn2set bug discovered by Jim McClune.

1.38  Fix bWeylSq dependency
1.37  Fix grdef symmetry bug
1.36  Add grcalcd(), grOptionqloadPath and allow makeg to not save a metric
1.35  Add grOptionMemory to grcalc and gralterGuts
1.34  Fix C(dn,dn,dn,dn) dependancy

==============================================================================


extras/maple.el
;;                 Maple editing support package
;;    (borrows heavily from modula2.el, ada.el and tex-mode.el)
;;
;;               Bruno Salvy (Bruno.Salvy@inria.fr)
;;
;; Major mode for Maple editing. It provides functions for editing
;; Maple code and interacting with Maple and Mint. See the documentation
;; of maple-mode.
;;
;; LCD Archive Entry:
;; maple-mode|Bruno Salvy|Bruno.Salvy@inria.fr|
;; Maple editing support|
;; 29-Sep-94|1.1|~/modes/maple.el.Z|
;;
;; modified 03/05/90: C-p and C-n in maple-shell. B.S.
;; modified 29/05/90: maple-prompt-regexp B.S.
;; modified 21/08/90: mint-level, M-p and M-n in maple-shell B.S.
;; modified 16/11/90: C-cT bound to maple-toggle-previous-next-behavior B.S.
;; modified 01/01/91: C-cC-l available also in maple-shell B.S.
;; modified 30/07/91: maple-help adapted to MapleV. B.S.
;; modified 09/03/92: nicer tabulation handling B.S.
;; modified 09/06/92: comment-start regexp B.S.

(provide 'maple)

(defvar maple-mode-syntax-table nil
  "Syntax table in use in maple-mode buffers.")

(if maple-mode-syntax-table
    ()
  (let ((table (make-syntax-table)))
    (modify-syntax-entry ?\\ "\\   " table)
    (modify-syntax-entry ?\n ">   "  table)
    (modify-syntax-entry ?\f ">   "  table)
    (modify-syntax-entry ?#  "<   "  table)
    (modify-syntax-entry ?\t "    "  table)
    (modify-syntax-entry ?*  ".   "  table)
    (modify-syntax-entry ?+  ".   "  table)
    (modify-syntax-entry ?-  ".   "  table)
    (modify-syntax-entry ?/  ".   "  table)
    (modify-syntax-entry ?=  ".   "  table)
    (modify-syntax-entry ?<  ".   "  table)
    (modify-syntax-entry ?>  ".   "  table)
    (modify-syntax-entry ?\' "\"   " table)
    (modify-syntax-entry ?\` "\"   " table)
    (modify-syntax-entry ?\" ".   "  table)
    (modify-syntax-entry ?\{ "(}  "  table)
    (modify-syntax-entry ?\[ "(]  "  table)
    (modify-syntax-entry ?\( "()  "  table)
    (modify-syntax-entry ?\} "){  "  table)
    (modify-syntax-entry ?\] ")[  "  table)
    (modify-syntax-entry ?\) ")(  "  table)
    (setq maple-mode-syntax-table table)))

(defvar maple-mode-map nil
  "Keymap used in Maple mode.")

(if maple-mode-map ()
  (let ((map (make-sparse-keymap)))
    (define-key map "\C-m" 'maple-newline)
    (define-key map "\C-?" 'maple-untab)
    (define-key map "\C-i" 'maple-tab)
    (define-key map "\C-c<" 'maple-backward-to-same-indent)
    (define-key map "\C-c>" 'maple-forward-to-same-indent)
    (define-key map "\C-c#" 'maple-inline-comment)
    (define-key map "\C-ce" 'maple-else)
    (define-key map "\C-cf" 'maple-for)
    (define-key map "\C-ch" 'maple-help)
    (define-key map "\C-ci" 'maple-if)
    (define-key map "\C-cl" 'maple-local)
    (define-key map "\C-cm" 'maple-modify)
    (define-key map "\C-cp" 'maple-procedure)
    (define-key map "\C-ct" 'maple-title)
    (define-key map "\C-cw" 'maple-while)
    (define-key map "\C-c\C-i" 'maple-make-library)
    (define-key map "\C-c\C-r" 'maple-region)
    (define-key map "\C-c\C-b" 'maple-buffer)
    (define-key map "\C-c\C-m" 'mint-buffer)
    (define-key map "\C-c\C-z" 'suspend-emacs)
    (setq maple-mode-map map)))

(defvar maple-indent 2 "*This variable gives the indentation in Maple-Mode")

(defun maple-mode ()
  "This is a mode intended to support program development in Maple.
All control constructs of Maple can be reached by typing
Control-C followed by the first character of the construct.
Use \\[maple-region] to run Maple on the current region under
a special subshell.  \\[maple-buffer] does the whole buffer,
\\[maple-make-library] doeis it with maple -s option.
\\[mint-buffer] runs mint on the buffer.

  Control-c f for           Control-c e else
  Control-c i if            Control-c l local
  Control-c w while         Control-c p proc
  Control-c # comment       Control-c h help
  Control-c m modify        Control-c t title
  Control-c ( paired parens
  Control-c Control-z suspend-emacs

  C-c < and C-c > move backward and forward respectively to the next line
  having the same (or lesser) level of indentation.

   maple-indent controls the number of spaces for each indentation.

   Note that in a Maple subshell buffer, C-p and C-n are bound to
   maple-previous-command and maple-next-commmand, while M-p and M-n
   are bound to previous-line and next-line. Use C-cT to toggle this
   behavior.

\\{maple-mode-map}
"
  (interactive)
  (kill-all-local-variables)
  (maple-define-common-keys maple-mode-map)
  (use-local-map maple-mode-map)
  (setq major-mode 'maple-mode)
  (setq mode-name "Maple")
  (make-local-variable 'comment-column)
  (setq comment-column 41)
  (make-local-variable 'end-comment-column)
  (setq end-comment-column 72)
  (set-syntax-table maple-mode-syntax-table)
  (make-local-variable 'paragraph-start)
  (setq paragraph-start (concat "^$\\|" page-delimiter))
  (make-local-variable 'paragraph-separate)
  (setq paragraph-separate paragraph-start)
  (make-local-variable 'paragraph-ignore-fill-prefix)
  (setq paragraph-ignore-fill-prefix t)
;;  (make-local-variable 'indent-line-function)
;;  (setq indent-line-function 'c-indent-line)
  (make-local-variable 'require-final-newline)
  (setq require-final-newline t)
  (make-local-variable 'comment-start)
  (setq comment-start "#")
  (make-local-variable 'comment-end)
  (setq comment-end "\n")
  (make-local-variable 'comment-column)
  (setq comment-column 41)
  (make-local-variable 'comment-start-skip)
  (setq comment-start-skip "#+ *")
  (make-local-variable 'comment-indent-hook)
  (setq comment-indent-hook 'c-comment-indent)
  (make-local-variable 'parse-sexp-ignore-comments)
  (setq parse-sexp-ignore-comments nil)
  (run-hooks 'maple-mode-hook))

(defun maple-newline ()
  "Insert a newline and indent following line like previous line."
  (interactive)
  (let  ((hpos (current-indentation)))
    (if (< (current-column) (current-indentation))
      (newline) (newline) (indent-to hpos))))

(defun maple-tab ()
  "Indent to next tab stop."
  (interactive)
  (if (> (current-column) (current-indentation))
      (insert "\t")
    (back-to-indentation)
    (let ((ci (current-indentation)))
      (backward-delete-char-untabify ci)
      (indent-to (* (1+ (/ ci maple-indent)) maple-indent)))))

(defun maple-tabsize (s)
  "changes spacing used for indentation. Reads spacing from minibuffer."
  (interactive "new indentation spacing: ")
  (setq maple-indent s))

(defun maple-untab ()
  "Delete backwards to previous tab stop."
  (interactive)
  (backward-delete-char-untabify
   (let ((ind (current-indentation)))
     (if (and (= ind (current-column)) (>= ind maple-indent))
         maple-indent 1))))

(defun maple-go-to-this-indent (step indent-level)
  "Move point repeatedly by <step> lines till the current line
has given indent-level or less, or the start/end of the buffer is hit.
Ignore blank lines and comments."
  (while (and
          (zerop (forward-line step))
          (or (looking-at "^[   ]*$")
              (looking-at "^[   ]*#")
              (looking-at "^<<[A-Za-z0-9_]+>>")
              (looking-at "^[A-Za-z0-9_]+:")
              (> (current-indentation) indent-level)))
    nil))

(defun maple-backward-to-same-indent ()
  "Move point backwards to nearest line with same indentation or less.
If not found, point is left at top of buffer."
  (interactive)
  (maple-go-to-this-indent -1 (current-indentation))
  (back-to-indentation))

(defun maple-forward-to-same-indent ()
  "Move point forwards to nearest line with same indentation or less.
If not found, point is left at start of last line in buffer."
  (interactive)
  (maple-go-to-this-indent 1 (current-indentation))
  (back-to-indentation))

(defun maple-for ()
  "Build skeleton for-loop statment, prompting for the loop parameters."
  (interactive)
  (let ((for (read-string "var: ")))
    (if (string-equal for "")
        (let ((to (read-string "to: ")))
          (if (not (string-equal to ""))
              (insert " to " to)))
      (insert "for " for)
      (let ((in (read-string "in: ")))
        (if (not (string-equal in ""))
            (insert " in " in)
          (let ((from (read-string "from: ")))
            (if (not (string-equal from ""))
                (insert " from " from)))
          (let ((by (read-string "by: ")))
            (if (not (string-equal by ""))
                (insert " by " by)))
          (let ((to (read-string "to: ")))
            (if (not (string-equal to ""))
                (insert " to " to)))))))
  (let ((while (read-string "while: ")))
    (if (not (string-equal while ""))
        (insert " while " while)))
  (insert " do")
  (maple-newline)
  (maple-newline)
  (insert "od;")
  (end-of-line 0)
  (maple-tab))

(defun maple-while ()
  "Build skeleton while-loop statment, prompting for the loop parameters."
  (interactive)
  (insert "while " (read-string "cond: "))
  (insert " do")
  (maple-newline)
  (maple-newline)
  (insert "od;")
  (end-of-line 0)
  (maple-tab))

(defun maple-title ()
  "Insert a comment block containing the module title, author, etc."
  (interactive)
  (if (eq (point) (point-min))
      nil
    (set-mark (point))
    (goto-line 1))
  (insert "\n\n")
  (previous-line 2)
  (insert "##    -*-Maple-*-\n")
  (insert "##\n##    Title: \t")
  (insert (read-string "Title: "))
  (insert "\n##    Created:\t")
  (insert (current-time-string))
  (insert "\n##    Author: \t")
  (insert (user-full-name))
  (insert (concat "\n##\t\t<" (user-login-name) "@" (system-name) ">\n"))
  (insert "##\n## Description: ")
  (end-of-line nil))

(defun maple-modify ()
  "Insert a comment block containing the modification, author, etc."
  (interactive)
  (set-mark (point))
  (goto-line 1)
  (while (char-equal (char-after (point)) 35)
    (forward-line 1))                                        
  (insert "##\n##    Modified: \t")
  (insert (current-time-string))
  (insert "\n##    Author: \t")
  (insert (user-full-name))
  (insert "\n##    Modification: ")
  (insert (read-string "Modification: "))
  (insert "\n##\n"))

(defun maple-if ()
  "Insert skeleton if statment, prompting for <boolean-expression>."
  (interactive)
  (insert "if " (read-string "cond: ") " then")
  (maple-newline)
  (maple-newline)
  (insert "fi;")
  (end-of-line 0)
  (maple-tab))

(defun maple-else ()
  "Add an elif clause to an if statement, prompting for the condition.
   When no condition is given, put an else."
  (interactive)
  (maple-untab)
  (let ((condition (read-string "elif: ")))
    (if (not (string-equal condition ""))
        (insert "elif " condition " then")
      (insert "else")))
  (maple-newline)
  (maple-tab))

(defun maple-local ()
  "Add a new local variable, inserting the word local if necessary."
  (interactive)
  (save-excursion
    (set-mark (point))
    (while (or (> (current-indentation) 0)
               (looking-at "#")
               (looking-at "end")
               (looking-at "option"))
      (forward-line -1))
    (let ((first-time))
      (if (looking-at "local")
          (setq first-time nil)
        (forward-line 1)
        (insert "local ;\n")
        (forward-line -1)
        (setq first-time t))
      (search-forward ";")
      (backward-char)
      (let ((newvar (read-string "New variable: ")))
        (if first-time (insert newvar)
          (insert ", " newvar))))))

(defun maple-procedure ()
  (interactive)
  (let ((name (read-string "Name: " ))
        args)
    (insert name ":=proc (")
    (insert (read-string "Arguments: ") ")")
    (let ((options (read-string "Options: ")))
      (if (not (string-equal options ""))
          (progn (maple-newline)
                 (insert "options " options ";"))))
    (maple-newline)
    (maple-newline)
    (insert "end: # ")
    (insert name)
    (end-of-line 0)
    (maple-tab)))

(defun maple-paired-parens ()
  "Insert a pair of round parentheses, placing point between them."
  (interactive)
  (insert "()")
  (backward-char))

(defun maple-inline-comment ()
  "Start a comment after the end of the line, indented at least COMMENT-COLUMN.
If starting after END-COMMENT-COLUMN, start a new line."
  (interactive)
  (end-of-line)
  (if (> (current-column) end-comment-column) (newline))
  (if (< (current-column) comment-column) (indent-to comment-column))
  (insert "#  "))

(defun maple-display-comment ()
  "Inserts 3 comment lines, making a display comment."
  (interactive)
  (insert "#\n# \n#")
  (end-of-line 0))

(defvar maple-lib-directory "/usr/local/lib/mapleV/lib")

(defun maple-help ()
  "Like describe-function in lisp-mode, tries to guess which function is
interesting for the user and prompts for confirmation, then displays help.
This could be much better if we called Maple to find the file, but I do not
see how it could be as fast. This will only work with version 5 or higher."
  (interactive)
  (save-excursion
    (let ((orig (point))
          eow bow)
      (forward-word -1)
      (setq bow (point))
      (forward-word 1)
      (setq eow (point))
      (let* ((posfuncname (buffer-substring bow eow))
             (funcname (read-string
                        (if (string-equal posfuncname "")
                            "Help about: "
                          (concat "Help about [" posfuncname "]: "))))
             filename)
        (and (string-equal funcname "")
             (setq funcname posfuncname))
        (if (string-match "[a-zA-Z]*\\[" funcname)
            (setq filename
                  (concat maple-lib-directory "/help/"
                          (substring funcname 0 (- (match-end 0) 1))
                          "/text/"
                          (substring funcname (match-end 0)
                                     (- (length funcname) 1))
                          ".m"))
          (setq filename (concat maple-lib-directory "/help/text/"
                                 funcname ".m")))
        (or (file-exists-p filename)
            (error "Cannot find help for this function"))
        (with-output-to-temp-buffer "*Maple Help*"
          (buffer-flush-undo standard-output)
          (save-excursion
            (set-buffer standard-output)
            (insert-file-contents filename)
            (goto-char (point-min))
            (kill-line 4)
            (replace-regexp "\230\226\214" "\n")
            (goto-char (point-min))
            (replace-regexp "[^\n -~]" "")))))))

;;; Invoking Maple in an inferior shell.

(defvar maple-command "maple"
  "The command to run maple on a file.")

(defvar maple-args ""
  "Arguments passed to maple. For old versions of Maple, -q could be useful")

(defvar maple-prompt-regexp "^[^-> \n]*>+ +" "\
*Regexp defining the prompt in Maple sessions.")

(defvar mint-command "mint"
  "The command to run mint on a file.  The name of the file will be appended
to this string, separated by <.")

(defvar mint-level "2"
  "Sets the verbosity of Mint")

(defun maple-define-common-keys (keymap)
  "Define the keys that we want defined both in maple-mode
and in the maple-shell."
  (define-key keymap "\C-c\C-k" 'maple-kill-job)
  (define-key keymap "\C-c\C-l" 'maple-recenter-output-buffer)
  (define-key keymap "\C-ch" 'maple-help))

(defvar maple-shell-map nil
  "Keymap for the maple shell.  A shell-mode-map with a few additions")

(defvar maple-temp-directory "/tmp/"
  "*Directory in which to create temporary files.")

(defvar zap-file nil
  "Temporary file name used for text being sent as input to Maple.")

(defun maple ()
  "Run maple in a buffer, without shell. Exiting maple will kill the buffer.
The buffer is called *Maple*, and the program used comes from variable
maple-command (default maple). The buffer is put in shell-mode with
a maple-syntax-table."
  (interactive)
  (if (get-buffer "*Maple*")
      nil
    (maple-start-shell "*Maple*"))
  (switch-to-buffer "*Maple*"))

(defun maple-start-shell (name)
  (require 'shell)
  (save-excursion
    (set-process-sentinel
     (if (not (string-equal maple-args ""))
         (start-process "maple" name
                        (concat exec-directory "env")
                        "TERM=emacs"  ; so that the process may know it 
                                        ; is called by emacs
                        "PAGER=cat"   ; because of 4.3 help
                                        ;                                                                 "PAGER=emacsclient"   ; because of 4.3 help
                        "-"
                        maple-command
                        maple-args)
       (start-process "maple" name
                      (concat exec-directory "env")
                      "TERM=emacs"  ; so that the process may know it 
                                        ; is called by emacs
                      "PAGER=cat"   ; because of 4.3 help
                                        ;                                                               "PAGER=emacsclient"   ; because of 4.3 help
                      "-"
                      maple-command))
     'maple-process-sentinel)
    (set-buffer name)
    (shell-mode)
    (make-local-variable 'shell-prompt-pattern)
    (setq shell-prompt-pattern maple-prompt-regexp)
    (set-syntax-table maple-mode-syntax-table)
    (setq maple-shell-map (copy-keymap shell-mode-map))
    (maple-define-common-keys maple-shell-map)
    (define-key maple-shell-map "\M-p" 'previous-line)
    (define-key maple-shell-map "\M-n" 'next-line)
    (define-key maple-shell-map "\C-p" 'maple-previous-command)
    (define-key maple-shell-map "\C-n" 'maple-next-command)
    (define-key maple-shell-map "\C-cT" 'maple-toggle-previous-next-behavior)
    (use-local-map maple-shell-map)))

(defun maple-process-sentinel (proc mesg)
  (let ((stat (process-status proc))
        (name (process-buffer proc)))
    (cond ((eq stat 'run) nil)
          ((eq stat 'stop)
           (continue-process proc)
           (if (and (not (eq (process-status proc) 'run))
                    name
                    (save-excursion
                      (kill-buffer name)))))
          (t (if (not name) nil
               (save-excursion
                 (kill-buffer name)))))))

(defun maple-region (beg end)
  "Run Maple on the current region.  A temporary file (zap-file) is
written in directory maple-temp-directory, but Maple is run in the current
directory."
  (interactive "r")
  (let ((name (concat "*" (buffer-name) "-Maple*")))
    (or (get-buffer name)
        (maple-start-shell name))
    (if zap-file
        (if (file-exists-p zap-file)
            (delete-file zap-file))
      (setq zap-file
            (expand-file-name
             (concat maple-temp-directory
                     (make-temp-name "#mz")))))
    (save-excursion
      (save-restriction
        (widen)
        (goto-char (point-min))
        (write-region beg end zap-file t nil))
      (send-string (get-buffer-process name)
                   (concat "read \`" zap-file "\`;\n")))
    (sit-for 1)
    (maple-recenter-output-buffer nil)))

(defun mint-shell (name)
  (require 'shell)
  (save-excursion
    (set-process-sentinel
     (start-process "mint" "*Mint*" mint-command "-i" mint-level name)
     'mint-process-sentinel)
    (set-buffer "*Mint*")
    (shell-mode)))

(defun mint-process-sentinel (proc mesg)
  (let ((stat (process-status proc))
        (name (process-buffer proc)))
    (cond ((eq stat 'run) nil)
          ((eq stat 'stop)
           (continue-process proc))
          (t nil))))

(defun mint-region (beg end)
  "Run Mint on the current region.  A temporary file (zap-file) is
written in directory maple-temp-directory, but Mint is run in the current
directory."
  (interactive "r")
  (let ((name "*Mint*"))
    (if (get-buffer name)
        (save-excursion (kill-buffer name)))
    (if zap-file
        (if (file-exists-p zap-file)
            (delete-file zap-file))
      (setq zap-file
            (expand-file-name
             (concat maple-temp-directory
                     (make-temp-name "#mz")))))
    (save-excursion
      (save-restriction
        (widen)
        (goto-char (point-min))
        (write-region beg end zap-file t nil))
      (mint-shell zap-file))
    (pop-to-buffer "*Mint*")))

(defun maple-buffer ()
  "Run Maple on current buffer.  See \\[maple-region] for more information."
  (interactive)
  (maple-region (point-min) (point-max)))

(defun maple-make-library ()
  "Run Maple on current buffer, with option -s."
  (interactive)
  (let ((maple-args "-s"))
    (maple-region (point-min) (point-max))))

(defun mint-buffer ()
  "Run Mint on current buffer.  See \\[mint-region] for more information."
  (interactive)
  (mint-region (point-min) (point-max)))

(defun maple-kill-job ()
  "Kill the currently running Maple job."
  (interactive)
  (if (get-buffer-process (current-buffer))
      (kill-buffer (current-buffer))
    (let ((name (concat "*" (buffer-name) "-Maple*")))
      (and (get-buffer name)
           (kill-buffer name)))))

(defun maple-recenter-output-buffer (linenum)
  "Redisplay buffer of Maple job output so that most recent output can be seen.
The last line of the buffer is displayed on
line LINE of the window, or centered if LINE is nil."
  (interactive "P")
  (let ((old-buffer (current-buffer))
        (maple-shell(if (get-buffer-process (current-buffer))
                        (current-buffer)
                      (or (get-buffer (concat "*" (buffer-name) "-Maple*"))
                          (get-buffer "*Maple*")))))
    (if (null maple-shell)
        (message "No Maple output buffer")
      (pop-to-buffer maple-shell)
      (goto-char (point-max))
      (recenter (if linenum
                    (prefix-numeric-value linenum)
                  (/(window-height) 2)))
      (pop-to-buffer old-buffer))))

(defun maple-previous-command ()
  "Recall previous maple command."
  (interactive)
  (maple-relative-command -1))

(defun maple-next-command ()
  "Step to maple next command line."
  (interactive)
  (maple-relative-command 1))

(defun maple-relative-command (dir)
  "Step to previous or next command line according to the first argument
being 1 or -1."
  (while (and (zerop (forward-line dir))
              (not (looking-at maple-prompt-regexp))
              (looking-at "^"))); forward-line at the end of a buffer
  (end-of-line))

(defun maple-toggle-previous-next-behavior ()
  "Change C-p/M-p C-n/M-n from previous-line and next-line to
maple-previous-command and maple-next-command and reciprocally"
  (interactive)
  (if (equal (key-binding "\C-p") 'previous-line)
      (progn
        (define-key maple-shell-map "\M-p" 'previous-line)
        (define-key maple-shell-map "\M-n" 'next-line)
        (define-key maple-shell-map "\C-p" 'maple-previous-command)
        (define-key maple-shell-map "\C-n" 'maple-next-command))
    (define-key maple-shell-map "\C-p" 'previous-line)
    (define-key maple-shell-map "\C-n" 'next-line)
    (define-key maple-shell-map "\M-p" 'maple-previous-command)
    (define-key maple-shell-map "\M-n" 'maple-next-command)))

;; Hilit patterns for use with hilit19 (by Denis):

(hilit-set-mode-patterns
 'maple-mode
 '(("[#].*$" nil comment)
   ("`[^`\n]*`" nil string)
   ("[ \t\n]\\(for\\|while\\|do\\|in\\|from\\|to\\|by\\|and\\|or\\|od\\|if\\|elif\\|then\\|else\\|fi\\)[ :;\t\n]" nil keyword)
   ("[ \t:=0-9A-Aa-z_]*\\(proc[ \t\n0-9A-Za-z():;,_]*)[\n]+\\|end[ :;\n]\\)" nil defun)
   ("\\(local\\|global\\)" nil type)
   ("\\(RETURN\\|ERROR\\)" nil keyword)
   )
 nil 'case-insensitive)

